File: src/main/java/baekjoon/backtraking/No9663.java
package baekjoon.backtraking;

import baekjoon.template.BaekjoonTemplate;

import java.io.IOException;

public class No9663 extends BaekjoonTemplate {
	static int count = 0;

	public static void main(String[] args) throws IOException {
		boolean[] up = new boolean[15];
		boolean[] diagonal1 = new boolean[29];
		boolean[] diagonal2 = new boolean[29];
		int input = Integer.parseInt(br.readLine());
		recursion(0, input, up, diagonal1, diagonal2);
		System.out.println(count);
	}

	private static void recursion(int i, int input, boolean[] up, boolean[] diagonal1, boolean[] diagonal2) {
		if (i == input) {
			count++;
			return;
		}
		for (int j = 0; j < input; j++) {
			if (up[j] || diagonal1[j + i] || diagonal2[j - i + input - 1]) {
				continue;
			}
			up[j] = true;
			diagonal1[j + i] = true;
			diagonal2[j - i + input - 1] = true;
			recursion(i + 1, input, up, diagonal1, diagonal2);
			up[j] = false;
			diagonal1[j + i] = false;
			diagonal2[j - i + input - 1] = false;
		}
	}

	@Override
	public void solution() throws IOException {

	}
}

File: src/main/java/baekjoon/backtraking/No9663.kt
import java.io.BufferedReader
import java.io.InputStreamReader

fun main() {
    val br = BufferedReader(InputStreamReader(System.`in`))
    val N = br.readLine().toInt()
    val up = BooleanArray(15)
    val diagonal1 = BooleanArray(29)
    val diagonal2 = BooleanArray(29)
    backtracking(N, up, diagonal1, diagonal2)


}

fun backtracking(N: Int, up: BooleanArray, diagonal1: BooleanArray, diagonal2: BooleanArray) {
    var count = 0
    fun recursion(x: Int) {
        if (x == N) {
            count++
            return
        }
        for (i in 0 until N) {
            if (up[i] || diagonal1[x + i] || diagonal2[x - i + N - 1]) {
                continue
            }
            up[i] = true
            diagonal1[x + i] = true
            diagonal2[x - i + N - 1] = true
            recursion(x + 1)
            up[i] = false
            diagonal1[x + i] = false
            diagonal2[x - i + N - 1] = false
        }
    }
    recursion(0)
    println(count)
}
